File: src/openaec_reports/core/styles.py
"""Huisstijl definities — kleuren, fonts, spacing, paragraph styles."""

from __future__ import annotations

import functools
from dataclasses import dataclass
from typing import TYPE_CHECKING

from reportlab.lib.colors import HexColor
from reportlab.lib.enums import TA_CENTER, TA_LEFT
from reportlab.lib.styles import ParagraphStyle, StyleSheet1

from openaec_reports.core.fonts import get_font_name, register_fonts

if TYPE_CHECKING:
    from openaec_reports.core.brand import BrandConfig


@functools.cache
def _ensure_fonts_registered() -> None:
    """Registreer fonts eenmalig (lazy, bij eerste gebruik)."""
    register_fonts()


# ============================================================
# 3BM Huisstijl Kleuren
# ============================================================


@dataclass(frozen=True)
class Colors:
    """3BM Huisstijl kleurenpalet."""

    primary: str = "#40124A"  # Donkerpaars (3BM huisstijl)
    secondary: str = "#38BDA0"  # Turquoise (3BM huisstijl)
    accent: str = "#2ECC71"  # Groen (voldoet)
    warning: str = "#E74C3C"  # Rood (voldoet niet)
    text: str = "#45243D"  # Donkerpaars tekst
    text_accent: str = "#56B49B"  # H2/H3 headings, TOC level 1
    text_light: str = "#7F8C8D"  # Lichtgrijs
    background: str = "#FFFFFF"  # Wit
    background_alt: str = "#F8F9FA"  # Lichtgrijs achtergrond
    rule: str = "#BDC3C7"  # Lijn kleur
    table_header_bg: str = "#45233C"  # Tabel header achtergrond
    table_header_text: str = "#FFFFFF"  # Tabel header tekst
    table_footer_bg: str = "#55B49B"  # Tabel footer/totaal rij
    separator: str = "#E0D0E8"  # Scheidingslijnen

    def as_hex(self, name: str) -> HexColor:
        """Retourneer kleur als ReportLab HexColor.

        Args:
            name: Kleurnaam (bijv. 'primary', 'text_light').
        """
        return HexColor(getattr(self, name))

    # Alias voor backward compatibility
    hex = as_hex


BM_COLORS = Colors()

# Gedeelde padding constante voor content block componenten (in points)
BLOCK_PADDING = 6


# ============================================================
# Font Configuratie
# ============================================================


@dataclass(frozen=True)
class FontConfig:
    """Font configuratie.

    Gebruikt Gotham fonts als ze beschikbaar zijn in assets/fonts/.
    Valt automatisch terug op Helvetica als Gotham niet geïnstalleerd is.
    """

    heading: str = "GothamBold"  # Fallback: Helvetica-Bold
    body: str = "GothamBook"  # Fallback: Helvetica
    medium: str = "GothamMedium"  # Fallback: Helvetica
    italic: str = "GothamBookItalic"  # Fallback: Helvetica-Oblique
    mono: str = "Courier"
    body_size: float = 9.5
    heading1_size: float = 18.0
    heading2_size: float = 13.0
    heading3_size: float = 11.0
    caption_size: float = 8.0
    footer_size: float = 7.5


def _make_font_config() -> FontConfig:
    """Maak FontConfig met effectieve font namen (Gotham of Helvetica fallback)."""
    _ensure_fonts_registered()
    return FontConfig(
        heading=get_font_name("GothamBold"),
        body=get_font_name("GothamBook"),
        medium=get_font_name("GothamMedium"),
        italic=get_font_name("GothamBookItalic"),
    )


BM_FONTS = _make_font_config()


# ============================================================
# Stylesheet
# ============================================================


def create_stylesheet(brand: BrandConfig | None = None) -> StyleSheet1:
    """Maak de 3BM stylesheet, optioneel met brand-specifieke overrides.

    Args:
        brand: Optionele brand configuratie met style overrides.

    Returns:
        StyleSheet1 met alle paragraph styles.
    """
    styles = StyleSheet1()

    styles.add(
        ParagraphStyle(
            name="Normal",
            fontName=BM_FONTS.body,
            fontSize=BM_FONTS.body_size,
            leading=BM_FONTS.body_size * 1.4,
            textColor=HexColor(BM_COLORS.text),
            alignment=TA_LEFT,
            spaceAfter=4,
        )
    )

    styles.add(
        ParagraphStyle(
            name="Heading1",
            parent=styles["Normal"],
            fontName=BM_FONTS.body,  # GothamBook, NIET Bold
            fontSize=BM_FONTS.heading1_size,
            leading=BM_FONTS.heading1_size * 1.3,
            textColor=HexColor(BM_COLORS.text),  # text kleur, NIET primary
            spaceBefore=12,
            spaceAfter=6,
        )
    )

    styles.add(
        ParagraphStyle(
            name="Heading2",
            parent=styles["Normal"],
            fontName=BM_FONTS.body,  # GothamBook
            fontSize=BM_FONTS.heading2_size,
            leading=BM_FONTS.heading2_size * 1.3,
            textColor=HexColor(BM_COLORS.text_accent),  # turquoise
            spaceBefore=10,
            spaceAfter=4,
        )
    )

    styles.add(
        ParagraphStyle(
            name="Heading3",
            parent=styles["Normal"],
            fontName=BM_FONTS.body,  # GothamBook
            fontSize=BM_FONTS.heading3_size,
            leading=BM_FONTS.heading3_size * 1.3,
            textColor=HexColor(BM_COLORS.text_accent),  # turquoise
            spaceBefore=8,
            spaceAfter=3,
        )
    )

    styles.add(
        ParagraphStyle(
            name="Caption",
            parent=styles["Normal"],
            fontSize=BM_FONTS.caption_size,
            textColor=HexColor(BM_COLORS.text_light),
            alignment=TA_CENTER,
        )
    )

    styles.add(
        ParagraphStyle(
            name="Footer",
            parent=styles["Normal"],
            fontSize=BM_FONTS.footer_size,
            textColor=HexColor(BM_COLORS.text_light),
        )
    )

    styles.add(
        ParagraphStyle(
            name="CoverTitle",
            parent=styles["Normal"],
            fontName=BM_FONTS.heading,
            fontSize=28,
            leading=34,
            textColor=HexColor(BM_COLORS.primary),
            alignment=TA_LEFT,
        )
    )

    styles.add(
        ParagraphStyle(
            name="CoverSubtitle",
            parent=styles["Normal"],
            fontName=BM_FONTS.body,
            fontSize=14,
            leading=18,
            textColor=HexColor(BM_COLORS.secondary),
            alignment=TA_LEFT,
        )
    )

    # Brand style overrides toepassen
    if brand and brand.styles:
        for style_name, overrides in brand.styles.items():
            if style_name in styles.byName:
                style = styles[style_name]
                for attr, value in overrides.items():
                    if attr == "textColor":
                        value = HexColor(value)
                    elif attr == "fontName":
                        value = get_font_name(value)
                    setattr(style, attr, value)

    return styles


BM_STYLES = create_stylesheet()


# ============================================================
# Gedeelde block styles (calculation / check_block)
# ============================================================


def block_style_reference() -> ParagraphStyle:
    """Referentie-stijl: klein, rechts uitgelijnd, lichtgrijs."""
    from reportlab.lib.enums import TA_RIGHT

    return ParagraphStyle(
        "_block_ref",
        parent=BM_STYLES["Normal"],
        fontSize=BM_FONTS.caption_size,
        leading=BM_FONTS.caption_size * 1.3,
        textColor=HexColor(BM_COLORS.text_light),
        alignment=TA_RIGHT,
        spaceAfter=0,
    )


def block_style_heading(text_color: str | None = None) -> ParagraphStyle:
    """Heading-stijl: bold, body size."""
    return ParagraphStyle(
        "_block_heading",
        parent=BM_STYLES["Normal"],
        fontName=BM_FONTS.heading,
        fontSize=BM_FONTS.body_size,
        leading=BM_FONTS.body_size * 1.3,
        textColor=HexColor(text_color or BM_COLORS.primary),
        spaceAfter=0,
    )


def block_style_body() -> ParagraphStyle:
    """Body-stijl: normaal, iets meer leading."""
    return ParagraphStyle(
        "_block_body",
        parent=BM_STYLES["Normal"],
        fontSize=BM_FONTS.body_size,
        leading=BM_FONTS.body_size * 1.4,
        spaceAfter=0,
    )


def block_style_mono() -> ParagraphStyle:
    """Monospace-stijl: Courier, body size."""
    return ParagraphStyle(
        "_block_mono",
        parent=BM_STYLES["Normal"],
        fontName=BM_FONTS.mono,
        fontSize=BM_FONTS.body_size,
        leading=BM_FONTS.body_size * 1.4,
        spaceAfter=0,
    )


def block_style_result(text_color: str | None = None) -> ParagraphStyle:
    """Resultaat-stijl: bold, iets groter, opvallend."""
    return ParagraphStyle(
        "_block_result",
        parent=BM_STYLES["Normal"],
        fontName=BM_FONTS.heading,
        fontSize=BM_FONTS.body_size + 1,
        leading=(BM_FONTS.body_size + 1) * 1.3,
        textColor=HexColor(text_color or BM_COLORS.primary),
        spaceAfter=0,
    )

File: rust/crates/openaec-core/src/styles.rs
//! Style system — colors, fonts, spacing, paragraph styles.
//!
//! Port of `core/styles.py`. Provides `StyleSet` as an explicit parameter
//! (no global mutable state). Produces `openaec_layout::ParagraphStyle` objects.

use openaec_layout::{Color, ParagraphStyle, Pt};
use serde::{Deserialize, Serialize};

use crate::brand::BrandConfig;

/// Shared padding constant for content block components (in points).
pub const BLOCK_PADDING: f64 = 6.0;

// ── Colors ─────────────────────────────────────────────────────────────

/// OpenAEC brand color palette.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Colors {
    pub primary: String,
    pub secondary: String,
    pub accent: String,
    pub warning: String,
    pub text: String,
    pub text_accent: String,
    pub text_light: String,
    pub background: String,
    pub background_alt: String,
    pub rule: String,
    pub table_header_bg: String,
    pub table_header_text: String,
    pub table_footer_bg: String,
    pub separator: String,
}

impl Default for Colors {
    fn default() -> Self {
        Self {
            primary: "#40124A".to_string(),
            secondary: "#38BDA0".to_string(),
            accent: "#2ECC71".to_string(),
            warning: "#E74C3C".to_string(),
            text: "#45243D".to_string(),
            text_accent: "#56B49B".to_string(),
            text_light: "#7F8C8D".to_string(),
            background: "#FFFFFF".to_string(),
            background_alt: "#F8F9FA".to_string(),
            rule: "#BDC3C7".to_string(),
            table_header_bg: "#45233C".to_string(),
            table_header_text: "#FFFFFF".to_string(),
            table_footer_bg: "#55B49B".to_string(),
            separator: "#E0D0E8".to_string(),
        }
    }
}

impl Colors {
    /// Resolve a color name to a `Color`.
    pub fn resolve(&self, name: &str) -> Option<Color> {
        let hex = match name {
            "primary" => &self.primary,
            "secondary" => &self.secondary,
            "accent" => &self.accent,
            "warning" => &self.warning,
            "text" => &self.text,
            "text_accent" => &self.text_accent,
            "text_light" => &self.text_light,
            "background" => &self.background,
            "background_alt" => &self.background_alt,
            "rule" => &self.rule,
            "table_header_bg" => &self.table_header_bg,
            "table_header_text" => &self.table_header_text,
            "table_footer_bg" => &self.table_footer_bg,
            "separator" => &self.separator,
            other if other.starts_with('#') => other,
            _ => return None,
        };
        Color::from_hex(hex)
    }
}

// ── Font configuration ─────────────────────────────────────────────────

/// Font role configuration.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FontConfig {
    pub heading: String,
    pub body: String,
    pub medium: String,
    pub italic: String,
    pub mono: String,
    pub body_size: f64,
    pub heading1_size: f64,
    pub heading2_size: f64,
    pub heading3_size: f64,
    pub caption_size: f64,
    pub footer_size: f64,
}

impl Default for FontConfig {
    fn default() -> Self {
        Self {
            heading: "Inter-Bold".to_string(),
            body: "Inter-Regular".to_string(),
            medium: "Inter-Medium".to_string(),
            italic: "Inter-RegularItalic".to_string(),
            mono: "Courier".to_string(),
            body_size: 9.5,
            heading1_size: 18.0,
            heading2_size: 13.0,
            heading3_size: 11.0,
            caption_size: 8.0,
            footer_size: 7.5,
        }
    }
}

// ── StyleSet ───────────────────────────────────────────────────────────

/// Complete style set — colors, fonts, and derived paragraph styles.
///
/// Passed explicitly as a parameter (no global state). Created from a
/// `BrandConfig` or using OpenAEC defaults.
#[derive(Debug, Clone)]
pub struct StyleSet {
    pub colors: Colors,
    pub fonts: FontConfig,
}

impl Default for StyleSet {
    fn default() -> Self {
        Self::default_openaec()
    }
}

impl StyleSet {
    /// Create a `StyleSet` with OpenAEC default colors and fonts.
    pub fn default_openaec() -> Self {
        Self {
            colors: Colors::default(),
            fonts: FontConfig::default(),
        }
    }

    /// Create a `StyleSet` from a `BrandConfig`.
    pub fn from_brand(brand: &BrandConfig) -> Self {
        let mut colors = Colors::default();

        // Apply brand colors
        if let Some(v) = brand.colors.get("primary") {
            colors.primary = v.clone();
        }
        if let Some(v) = brand.colors.get("secondary") {
            colors.secondary = v.clone();
        }
        if let Some(v) = brand.colors.get("accent") {
            colors.accent = v.clone();
        }
        if let Some(v) = brand.colors.get("warning") {
            colors.warning = v.clone();
        }
        if let Some(v) = brand.colors.get("text") {
            colors.text = v.clone();
        }
        if let Some(v) = brand.colors.get("text_accent") {
            colors.text_accent = v.clone();
        }
        if let Some(v) = brand.colors.get("text_light") {
            colors.text_light = v.clone();
        }
        if let Some(v) = brand.colors.get("background") {
            colors.background = v.clone();
        }
        if let Some(v) = brand.colors.get("background_alt") {
            colors.background_alt = v.clone();
        }
        if let Some(v) = brand.colors.get("table_header_bg") {
            colors.table_header_bg = v.clone();
        }
        if let Some(v) = brand.colors.get("table_header_text") {
            colors.table_header_text = v.clone();
        }
        if let Some(v) = brand.colors.get("table_footer_bg") {
            colors.table_footer_bg = v.clone();
        }
        if let Some(v) = brand.colors.get("separator") {
            colors.separator = v.clone();
        }

        // Apply brand fonts
        let mut fonts = FontConfig::default();
        if let Some(v) = brand.fonts.get("heading") {
            fonts.heading = v.clone();
        }
        if let Some(v) = brand.fonts.get("body") {
            fonts.body = v.clone();
        }
        if let Some(v) = brand.fonts.get("medium") {
            fonts.medium = v.clone();
        }
        if let Some(v) = brand.fonts.get("italic") {
            fonts.italic = v.clone();
        }
        if let Some(v) = brand.fonts.get("mono") {
            fonts.mono = v.clone();
        }

        Self { colors, fonts }
    }

    // ── Paragraph style factories ──────────────────────────────────────

    /// Normal body text style.
    pub fn normal(&self) -> ParagraphStyle {
        ParagraphStyle {
            font_size: Pt(self.fonts.body_size as f32),
            leading: Pt((self.fonts.body_size * 1.4) as f32),
            text_color: self.color_or_default(&self.colors.text, Color::BLACK),
            space_after: Pt(4.0),
            ..Default::default()
        }
    }

    /// Heading 1 style.
    pub fn heading1(&self) -> ParagraphStyle {
        ParagraphStyle {
            font_size: Pt(self.fonts.heading1_size as f32),
            leading: Pt((self.fonts.heading1_size * 1.3) as f32),
            text_color: self.color_or_default(&self.colors.text, Color::BLACK),
            bold: false,
            space_before: Pt(12.0),
            space_after: Pt(6.0),
            ..Default::default()
        }
    }

    /// Heading 2 style.
    pub fn heading2(&self) -> ParagraphStyle {
        ParagraphStyle {
            font_size: Pt(self.fonts.heading2_size as f32),
            leading: Pt((self.fonts.heading2_size * 1.3) as f32),
            text_color: self.color_or_default(&self.colors.text_accent, Color::BLACK),
            bold: false,
            space_before: Pt(10.0),
            space_after: Pt(4.0),
            ..Default::default()
        }
    }

    /// Heading 3 style.
    pub fn heading3(&self) -> ParagraphStyle {
        ParagraphStyle {
            font_size: Pt(self.fonts.heading3_size as f32),
            leading: Pt((self.fonts.heading3_size * 1.3) as f32),
            text_color: self.color_or_default(&self.colors.text_accent, Color::BLACK),
            bold: false,
            space_before: Pt(8.0),
            space_after: Pt(3.0),
            ..Default::default()
        }
    }

    /// Caption style — small, centered, light grey.
    pub fn caption(&self) -> ParagraphStyle {
        ParagraphStyle {
            font_size: Pt(self.fonts.caption_size as f32),
            leading: Pt((self.fonts.caption_size * 1.3) as f32),
            text_color: self.color_or_default(&self.colors.text_light, Color::GREY),
            ..Default::default()
        }
    }

    /// Footer style.
    pub fn footer(&self) -> ParagraphStyle {
        ParagraphStyle {
            font_size: Pt(self.fonts.footer_size as f32),
            leading: Pt((self.fonts.footer_size * 1.3) as f32),
            text_color: self.color_or_default(&self.colors.text_light, Color::GREY),
            ..Default::default()
        }
    }

    /// Cover title style.
    pub fn cover_title(&self) -> ParagraphStyle {
        ParagraphStyle {
            font_size: Pt(28.0),
            leading: Pt(34.0),
            bold: true,
            text_color: self.color_or_default(&self.colors.primary, Color::BLACK),
            ..Default::default()
        }
    }

    /// Cover subtitle style.
    pub fn cover_subtitle(&self) -> ParagraphStyle {
        ParagraphStyle {
            font_size: Pt(14.0),
            leading: Pt(18.0),
            text_color: self.color_or_default(&self.colors.secondary, Color::BLACK),
            ..Default::default()
        }
    }

    // ── Block-specific styles ──────────────────────────────────────────

    /// Reference style — small, right-aligned, light grey.
    pub fn block_reference(&self) -> ParagraphStyle {
        ParagraphStyle {
            font_size: Pt(self.fonts.caption_size as f32),
            leading: Pt((self.fonts.caption_size * 1.3) as f32),
            text_color: self.color_or_default(&self.colors.text_light, Color::GREY),
            space_after: Pt(0.0),
            ..Default::default()
        }
    }

    /// Block heading style — bold, body size, primary color.
    pub fn block_heading(&self, text_color: Option<&str>) -> ParagraphStyle {
        let color = text_color
            .and_then(Color::from_hex)
            .unwrap_or_else(|| {
                self.color_or_default(&self.colors.primary, Color::BLACK)
            });
        ParagraphStyle {
            font_size: Pt(self.fonts.body_size as f32),
            leading: Pt((self.fonts.body_size * 1.3) as f32),
            bold: true,
            text_color: color,
            space_after: Pt(0.0),
            ..Default::default()
        }
    }

    /// Block body style — normal, slightly more leading.
    pub fn block_body(&self) -> ParagraphStyle {
        ParagraphStyle {
            font_size: Pt(self.fonts.body_size as f32),
            leading: Pt((self.fonts.body_size * 1.4) as f32),
            text_color: self.color_or_default(&self.colors.text, Color::BLACK),
            space_after: Pt(0.0),
            ..Default::default()
        }
    }

    /// Block monospace style — Courier, body size.
    pub fn block_mono(&self) -> ParagraphStyle {
        ParagraphStyle {
            font_size: Pt(self.fonts.body_size as f32),
            leading: Pt((self.fonts.body_size * 1.4) as f32),
            text_color: self.color_or_default(&self.colors.text, Color::BLACK),
            space_after: Pt(0.0),
            ..Default::default()
        }
    }

    /// Block result style — bold, slightly larger, primary color.
    pub fn block_result(&self, text_color: Option<&str>) -> ParagraphStyle {
        let color = text_color
            .and_then(Color::from_hex)
            .unwrap_or_else(|| {
                self.color_or_default(&self.colors.primary, Color::BLACK)
            });
        let size = self.fonts.body_size + 1.0;
        ParagraphStyle {
            font_size: Pt(size as f32),
            leading: Pt((size * 1.3) as f32),
            bold: true,
            text_color: color,
            space_after: Pt(0.0),
            ..Default::default()
        }
    }

    // ── Helpers ─────────────────────────────────────────────────────────

    fn color_or_default(&self, hex: &str, fallback: Color) -> Color {
        Color::from_hex(hex).unwrap_or(fallback)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_default_colors() {
        let c = Colors::default();
        assert_eq!(c.primary, "#40124A");
        assert_eq!(c.secondary, "#38BDA0");
        assert_eq!(c.text, "#45243D");
    }

    #[test]
    fn test_color_resolve() {
        let c = Colors::default();
        assert!(c.resolve("primary").is_some());
        assert!(c.resolve("text").is_some());
        assert!(c.resolve("#FF0000").is_some());
        assert!(c.resolve("nonexistent").is_none());
    }

    #[test]
    fn test_default_fonts() {
        let f = FontConfig::default();
        assert_eq!(f.heading, "Inter-Bold");
        assert_eq!(f.body, "Inter-Regular");
        assert_eq!(f.body_size, 9.5);
    }

    #[test]
    fn test_styleset_default() {
        let ss = StyleSet::default_openaec();
        assert_eq!(ss.colors.primary, "#40124A");
        assert_eq!(ss.fonts.heading, "Inter-Bold");
    }

    #[test]
    fn test_styleset_from_brand() {
        let mut brand = BrandConfig::default();
        brand
            .colors
            .insert("primary".to_string(), "#FF0000".to_string());
        brand
            .fonts
            .insert("heading".to_string(), "CustomBold".to_string());

        let ss = StyleSet::from_brand(&brand);
        assert_eq!(ss.colors.primary, "#FF0000");
        assert_eq!(ss.fonts.heading, "CustomBold");
        // Unchanged defaults
        assert_eq!(ss.colors.secondary, "#38BDA0");
    }

    #[test]
    fn test_normal_style() {
        let ss = StyleSet::default_openaec();
        let style = ss.normal();
        assert_eq!(style.font_size, Pt(9.5));
        assert_eq!(style.space_after, Pt(4.0));
    }

    #[test]
    fn test_heading_styles() {
        let ss = StyleSet::default_openaec();
        let h1 = ss.heading1();
        let h2 = ss.heading2();
        assert_eq!(h1.font_size, Pt(18.0));
        assert_eq!(h2.font_size, Pt(13.0));
    }

    #[test]
    fn test_block_heading_custom_color() {
        let ss = StyleSet::default_openaec();
        let style = ss.block_heading(Some("#FF0000"));
        // Should use the provided color
        assert_eq!(style.text_color, Color::rgb(255, 0, 0));
    }

    #[test]
    fn test_block_padding_constant() {
        assert_eq!(BLOCK_PADDING, 6.0);
    }
}
